File: Editor/MapEditor/land.cpp
/*
    RPG Paper Maker Copyright (C) 2017-2023 Wano

    RPG Paper Maker engine is under proprietary license.
    This source code is also copyrighted.

    Use Commercial edition for commercial use of your games.
    See RPG Paper Maker EULA here:
        http://rpg-paper-maker.com/index.php/eula.
*/

#include "land.h"
#include "qbox3d.h"
#include "floor.h"
#include "lands.h"

const QString LandDatas::JSON_UP = "up";
const QString LandDatas::JSON_TEXTURE = "t";

// -------------------------------------------------------
//
//  CONSTRUCTOR / DESTRUCTOR / GET / SET
//
// -------------------------------------------------------

LandDatas::LandDatas(QRect* texture, bool up) :
    MapElement(),
    m_textureRect(texture),
    m_up(up)
{

}

LandDatas::~LandDatas()
{
    delete m_textureRect;
}

bool LandDatas::operator==(const LandDatas& other) const {
    return MapElement::operator==(other) && m_up == other.m_up &&
           m_textureRect->x() == other.m_textureRect->x() &&
           m_textureRect->y() == other.m_textureRect->y() &&
           m_textureRect->width() == other.m_textureRect->width() &&
           m_textureRect->height() == other.m_textureRect->height();
}

bool LandDatas::operator!=(const LandDatas& other) const {
    return !operator==(other);
}

QRect *LandDatas::textureRect() const { return m_textureRect; }

bool LandDatas::up() const {
    return m_up;
}

MapEditorSubSelectionKind LandDatas::getSubKind() const{
    return MapEditorSubSelectionKind::None;
}

// -------------------------------------------------------
//
//  INTERMEDIARY FUNCTIONS
//
// -------------------------------------------------------

void LandDatas::initializeVertices(int, int, int, QVector<Vertex>&,
                                   QVector<GLuint>&, Position&, int&)
{

}

// -------------------------------------------------------

float LandDatas::intersection(int squareSize, QRay3D& ray, Position& position) {
    return staticIntersection(squareSize, ray, position, m_up);
}

// -------------------------------------------------------

float LandDatas::staticIntersection(int squareSize, QRay3D& ray, Position& position,
    bool up)
{
    QVector3D pos, size, center;
    getUpPosSizeCenter(pos, size, center, squareSize, position, up);

    QVector3D vecA = Lands::verticesQuad[0] * size + pos,
              vecC = Lands::verticesQuad[2] * size + pos;
    QBox3D box(vecA, vecC);

    return box.intersection(ray);
}

// -------------------------------------------------------

void LandDatas::getUpPosSizeCenter(QVector3D& pos, QVector3D& size, QVector3D&
    center, int squareSize, Position &position, bool up)
{
    // Position
    float yLayerOffset = position.layer() * 0.05f;
    if (!up)
    {
        yLayerOffset *= -1;
    }
    float yPosition = position.getY() + yLayerOffset;
    pos.setX(position.x() * squareSize);
    pos.setY(yPosition);
    pos.setZ(position.z() * squareSize);

    // Size
    size.setX(squareSize);
    size.setY(0.0f);
    size.setZ(squareSize);

    // Center
    center.setX(pos.x() + (squareSize / 2));
    center.setY(pos.y());
    center.setZ(pos.z() + (squareSize / 2));
}

// -------------------------------------------------------

void LandDatas::getPosSizeCenterLand(QVector3D& pos, QVector3D& size, QVector3D
    &center, int squareSize, Position &position)
{
    getUpPosSizeCenter(pos, size, center, squareSize, position, m_up);
}

// -------------------------------------------------------
//
//  READ / WRITE
//
// -------------------------------------------------------

void LandDatas::read(const QJsonObject &json){
    MapElement::read(json);

    if (json.contains(JSON_UP))
        m_up = json[JSON_UP].toBool();

    QJsonArray tab = json[JSON_TEXTURE].toArray();
    m_textureRect->setLeft(tab[0].toInt());
    m_textureRect->setTop(tab[1].toInt());

    if (tab.size() > 2) {
        m_textureRect->setWidth(tab[2].toInt());
        m_textureRect->setHeight(tab[3].toInt());
    }
    else {
        m_textureRect->setWidth(1);
        m_textureRect->setHeight(1);
    }
}

// -------------------------------------------------------

void LandDatas::write(QJsonObject &json) const{
    MapElement::write(json);

    if (!m_up)
        json[JSON_UP] = m_up;

    QJsonArray tab;
    tab.append(m_textureRect->left());
    tab.append(m_textureRect->top());
    if (m_textureRect->width() != 1 || m_textureRect->height() != 1) {
        tab.append(m_textureRect->width());
        tab.append(m_textureRect->height());
    }
    json[JSON_TEXTURE] = tab;
}

File: src/editor/mapElements/Land.ts
/*
    RPG Paper Maker Copyright (C) 2017-2024 Wano

    RPG Paper Maker engine is under proprietary license.
    This source code is also copyrighted.

    Use Commercial edition for commercial use of your games.
    See RPG Paper Maker EULA here:
        http://rpg-paper-maker.com/index.php/eula.
*/

import * as THREE from 'three';
import { MapElement, Scene } from '../Editor';
import { Base } from './Base';
import { CustomGeometry, Position, Project, Rectangle } from '../core';
import { BINDING, BindingType, ELEMENT_MAP_KIND } from '../common';

class Land extends Base {
	public static readonly JSON_UP = 'up';
	public static readonly JSON_TEXTURE = 't';
	public static readonly DEFAULT_UP = true;

	public up!: boolean;
	public texture!: Rectangle;

	public static readonly bindings: BindingType[] = [
		['up', 'up', true, BINDING.BOOLEAN],
		['texture', 't', undefined, BINDING.RECTANGLE],
	];

	static getBindings(additionnalBinding: BindingType[]) {
		return [...Land.bindings, ...additionnalBinding];
	}

	static areLandsEquals(
		landBefore: Land | null,
		autotileAfter: number,
		textureAfter: Rectangle,
		kindAfter: ELEMENT_MAP_KIND
	) {
		if (landBefore === null) {
			return kindAfter === ELEMENT_MAP_KIND.NONE;
		} else {
			if (landBefore.kind === kindAfter) {
				if (landBefore instanceof MapElement.Autotile) {
					return landBefore.autotileID === autotileAfter && landBefore.texture.equals(textureAfter);
				} else {
					return landBefore.texture.equals(textureAfter);
				}
			} else {
				return false;
			}
		}
	}

	equals(mapElement: MapElement.Base) {
		if (mapElement.kind === this.kind) {
			const land = mapElement as MapElement.Land;
			return land.texture.equals(this.texture) && land.up === this.up;
		} else {
			return false;
		}
	}

	toString(): string {
		return 'LAND';
	}

	getIndex(width: number): number {
		return this.texture.x + this.texture.y * width;
	}

	updateGeometryLand(
		geometry: CustomGeometry,
		position: Position,
		width: number,
		height: number,
		x: number,
		y: number,
		w: number,
		h: number,
		count: number,
		squareWidth = 1,
		squareHeight = 1
	) {
		const localPosition = position.toVector3();
		const a = localPosition.x;
		let yLayerOffset = position.layer * Scene.Map.current!.camera.getYOffsetDepth();
		if (!this.up) {
			yLayerOffset *= -1;
		}
		const b = localPosition.y + yLayerOffset;
		const c = localPosition.z;

		// Vertices
		const vecA = new THREE.Vector3(a - Project.SQUARE_SIZE / 2, b, c - Project.SQUARE_SIZE / 2);
		const vecB = new THREE.Vector3(a + (Project.SQUARE_SIZE / 2) * squareWidth, b, c - Project.SQUARE_SIZE / 2);
		const vecC = new THREE.Vector3(
			a + (Project.SQUARE_SIZE / 2) * squareWidth,
			b,
			c + (Project.SQUARE_SIZE / 2) * squareHeight
		);
		const vecD = new THREE.Vector3(a - Project.SQUARE_SIZE / 2, b, c + (Project.SQUARE_SIZE / 2) * squareHeight);
		const center = new THREE.Vector3(a, b, c);
		Base.rotateQuadEuler(vecA, vecB, vecC, vecD, center, position.toRotationEuler());

		// Vertices
		geometry.pushQuadVertices(vecA, vecB, vecC, vecD);

		// Indices
		geometry.pushQuadIndices(count * 4, position);

		// UVs
		const coefX = Base.COEF_TEX / width;
		const coefY = Base.COEF_TEX / height;
		x += coefX;
		y += coefY;
		w -= coefX * 2;
		h -= coefY * 2;
		const texA = new THREE.Vector2();
		const texB = new THREE.Vector2();
		const texC = new THREE.Vector2();
		const texD = new THREE.Vector2();
		CustomGeometry.uvsQuadToTex(texA, texB, texC, texD, x, y, w, h);
		geometry.pushQuadUVs(texA, texB, texC, texD);
	}

	read(json: Record<string, any>, additionnalBinding: BindingType[] = []) {
		super.read(json, Land.getBindings(additionnalBinding));
	}

	write(json: Record<string, any>, additionnalBinding: BindingType[] = []) {
		super.write(json, Land.getBindings(additionnalBinding));
	}
}

export { Land };
